boards/BMS/Src/App/states/App_PreChargeState.c

```diff
@@ -61,12 +59,15 @@ static void PreChargeStateRunOnTick100Hz(struct StateMachine *const state_machin
         const bool is_ts_rising_quickly =
             (ts_voltage > threshold_voltage) && (elapsed_time <= PRECHARGE_COMPLETION_LOWER_BOUND);
         const bool is_charger_connected = App_Charger_IsConnected(charger);
-        const bool has_precharge_fault =
-            (is_charger_connected) ? is_ts_rising_slowly : (is_ts_rising_slowly | is_ts_rising_quickly);
+        const bool has_precharge_fault  = App_PrechargeRelay_CheckFaults(
+            can_tx, is_charger_connected, is_ts_rising_slowly, is_ts_rising_quickly, &precharge_fault_count);
 
         if (has_precharge_fault)
         {
-            App_SharedStateMachine_SetNextState(state_machine, App_GetFaultState());
+            if (precharge_fault_count >= 3)
+                App_SharedStateMachine_SetNextState(state_machine, App_GetFaultState());
+            else
+                App_SharedStateMachine_SetNextState(state_machine, App_GetInitState());
```

Comment:
Instead of checking if the fault count >= 3 here too, why doesn't App_PrechargeRelay_CheckFaults just return whether or not there is a precharge shutdown fault and then all the checking is hidden?

Reply:
Also I don't think this will work: If there's a precharge fault, before there's been enough time to count to 3, we'll first go to init state. So fault state here is unreachable via this logic.
Also you're resetting the fault count to zero at line 49, so it can't get to 3 ever anyway.

Reply:
> Also I don't think this will work: If there's a precharge fault, before there's been enough time to count to 3, we'll first go to init state. So fault state here is unreachable via this logic. Also you're resetting the fault count to zero at line 49, so it can't get to 3 ever anyway.

Ah good point. Setting the fault count as a static int should resolve it being reset, but I'm not sure I quite understand your first point. If the fault count does not reset, going back to Init state should not cause any issues should it? From @Eddy13013's comment [here](https://github.com/UBCFormulaElectric/Consolidated-Firmware/pull/848#discussion_r1013585046) it sounds like that is the intended behavior but if you have something else in mind I'm all ears

Reply:
> Instead of checking if the fault count >= 3 here too, why doesn't App_PrechargeRelay_CheckFaults just return whether or not there is a precharge shutdown fault and then all the checking is hidden?

I agree it's a bit cludgy, but unless we want to change to init state from inside App_PrechargeRelay_CheckFaults, we need some way to communicate to PrechargeState the difference between a "fault" and a "shutdown fault". I suppose rather than passing a pointer to precharge_fault_count I could just pass a pointer to a bool, which would probably be more along the lines of what you're suggesting?

Reply:
Yeah ignore the first point I thought this was still a debounce thing
Would it be a problem then setting the fault count to be zero if a fault is not set, in CheckFaults? Since if there's a fault, we got Init, then back to Precharge, and initially there won't be a fault so the counter will be reset. Probably you only want to zero the fault counter on a successful precharge, i.e. transition to drive state?
I'd also suggest moving all the fault counting to the state machine, then you don't need to pass the pointer, but that's up to you

Reply:
I agree with setting the counter to zero on a transition to drive state, as for implmenting the counter in CheckFaults vs Statemachine, I think that part of the goal of this transition is to make the state machine code more readable, i.e. more like pseudocode, and "hiding" the counter inside checkfaults might make the fault transitions easier to read. I'll look into the idea though, see what is the most intuitive to read, as that is the goal.

Reply:
@gtaharaedmonds Okay I made some major changes after consulting @Eddy13013, let me know how you feel about the new implementation